chore(dev): enforce code policies

allardy opened the pull request:
This PR brings some restrictions on code which will need to be fixed before the code can be committed. Eventually most of these rules will be moved to the general `tsconfig.json` file, but since we want to avoid refactoring the whole codebase with those rules from the beginning, they will only be enforced when you commit files to the repository.

** Most of these rules will be fixed automatically when the file is saved, once they are moved to `tslint.json` when the codebase is completely up-to-date with the rules

Here's a summary of what needs to be addressed before commiting the code: 

- Enforce ES6 shorthand notation for methods and property
- Use an interface instead of a type when a type is not required
- Avoid using backticks unless you really need it for a template literal
- Use string interpolation instead of concatenating string (ex: "text" + 2)
- If you have a class method that doesn't use `this`, put it outside of the class. Most of the time those are helper methods which can be put in a separate file
- This will give an error when trying to commit `console.log`
- It will also warn when awaiting a promise in the return statement
- Adds a strict equality check - need to use === (except for null/undefined, where `!= undefined` would match both)
- Warn when there are multiple imports from the same module on different lines (they must be merged)

The rule preventing the use of `null` is removed, so we don't have to disable the linter or use eval anymore.

These situations will give an error when trying to commit. If it's a partial commit and you need to keep those checks, there is a flag that can be used to bypass those verifications, but they must be addressed before the final commit.

The new rules are in `tslint.newrules.json` and you can get more informations on this page: https://palantir.github.io/tslint/rules/

Here's an example of the output when committing: 
![image](https://user-images.githubusercontent.com/42552874/88883229-cdeec900-d201-11ea-9922-3983730569c1.png)

allardy committed:
chore(dev): enforce code policies

allardy committed:
enable null and allow double quote for jsx

allardy committed:
fix unrelated typings

allardy committed:
updated typescript & quick fixes & add lint command

allardy commented:
I added two helper commands that can be run before committing: 
- `yarn run cmd lint` which runs the same check as the pre-commit hook
- `yarn run cmd lint --fix` which tries to auto-fix what it can fix

allardy committed:
added file limit check & ignore stash backup

allardy committed:
missing no stash

allardy commented:
Also added a check to prevent committing files larger than 30mb, maybe we can put this limit smaller.
@Ierezell what do you think?

spg commented:
Also, make sure all those new checks (and the file size check) is run on the CI

spg commented:
Personally, I prefer not having any pre-commit hooks that run linting. While developing, the linter can get annoying every time you commit. Most times, I just want to commit some code, not care about linting at this moment, and move on.

I prefer fixing linting issues at the last minute (when opening a PR).

So, I would remove the pre-commit hook altogether (this means we could get rid of Husky), and only run the checks on the CI. 

I know not everyone will agree with me, but I'm open to discussion.

allardy commented:
We used to have the pre-commit hook, but I removed it because it happened twice that I lost like 2 hours of work. The problem was caused by the stash mechanism, which I removed here. 

I agree that some people commits a lot, what do you think of using a pre-push check instead ? I'm just used to fix linting errors on the go, it could be a nice habit to take to make changes as you go, instead of doing through a list of 40 changes when you open a PR. 

The reason I want to add the pre-commit now is that since it's a different set of rules, it will not show as errors in the CodeStyle CI of github

There is also the possibility to run the pre-commit hook with the flag --fix, so it would try to fix changes in your staged files as you commit, but you get no warning about that.

allardy committed:
moved checks to a pre-push hook

allardy committed:
test

allardy committed:
moving checks to CI

allardy committed:
remove ci check

allardy committed:
test change

allardy committed:
more test

allardy committed:
testing ci

allardy committed:
fix for cli

allardy committed:
test

allardy committed:
test

allardy committed:
test2

allardy committed:
lalala

allardy committed:
another try

allardy committed:
testtt

allardy committed:
testing

allardy committed:
cleanup

allardy committed:
remove test comment
